### PR TITLE
[core][autoscaler][v2] scheduler - gang requests schedulng [6/x]

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This PR implements the logic to handle gang resource requests with placement constraints.

## Related issue number

- https://github.com/ray-project/ray/issues/42840

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/scheduler.py [hunk 1]
```diff
@@ -318,7 +318,22 @@ def _try_schedule_one(self, request: ResourceRequest, is_constraint: bool) -> bo
             True if the resource request is scheduled on this node.
         """
 
-        # TODO: add labels when implementing the gang scheduling.
+        # Check if there's placement constraints that are not satisfied.
+        for constraint in request.placement_constraints:
+            if constraint.HasField("anti_affinity"):
+                anti_affinity = constraint.anti_affinity
+                if (
+                    anti_affinity.label_name in self.labels
```
Comment (scv119) [inline python/ray/autoscaler/v2/scheduler.py:326]: a little bit lost.. can you remind me how we convert pg into labels again?

**Review by scv119 — APPROVED** [review]
let's have @jjyao's review as well

File: python/ray/autoscaler/v2/scheduler.py (hunk 1 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/scheduler.py:326]: For a `pg=([CPU:1, CPU:1], strict_spread` we add anti-affinity constraints as `{ pg_<pg_id>: ""}`

Specifically, this 
https://github.com/ray-project/ray/blob/55838fc47bf5531d7c73c9b48064a6d1fd4f2d12/src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc#L145-L171

and this:

https://github.com/ray-project/ray/blob/55838fc47bf5531d7c73c9b48064a6d1fd4f2d12/src/ray/gcs/pb_util.h#L408-L429

File: python/ray/autoscaler/v2/scheduler.py [hunk 2]
```diff
@@ -468,7 +468,22 @@ def _try_schedule_one(
...
+                    and anti_affinity.label_value
+                    == self.labels[anti_affinity.label_name]
+                ):
+                    # The node already has a label that matches the anti-affinity
+                    return False
+
+            # We don't need to check for affinity constraints here since
+            # affinity doesn't affect if a request can be scheduled on a node.
+            # Affinity constraints are only used for scoring.
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:485]: Mention that the bundles are combined into one and STRICT_PACK doesn't have partial recovery.

File: python/ray/autoscaler/v2/scheduler.py [hunk 3]
```diff
@@ -482,8 +497,33 @@ def _try_schedule_one(
         # Add the request to the node.
         self.sched_requests[resource_request_source].append(request)
 
+        # Update the dynamic labels if there's any
+        for constraint in request.placement_constraints:
+            if constraint.HasField("affinity"):
+                affinity = constraint.affinity
+                self._add_label(affinity.label_name, affinity.label_value)
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:504]: Not needed

File: python/ray/autoscaler/v2/scheduler.py [hunk 4]
```diff
@@ -1137,7 +1177,42 @@ def _sched_gang_resource_requests(
...
-        return []
+
+        def _sort_gang_resource_requests(req: GangResourceRequest) -> Tuple:
+            """
+            Key function for sorting the gang resource request by:
+                1. the number of placement constraints in the gang request.
+                2. the number of resource requests in the gang request.
+            """
+            total_placement_constraints = 0
+            for rr in req.requests:
```
Comment (jjyao) [inline python/ray/autoscaler/v2/scheduler.py:1188]: rr?

File: python/ray/autoscaler/v2/utils.py [hunk 5]
```diff
@@ -157,8 +172,96 @@ def make(resources_map: Dict[str, float]) -> ResourceRequest:
...
+                    )
+                )
+            else:
+                raise ValueError(f"Unknown constraint type: {constraint_type}")
+
         return request
 
+    @staticmethod
+    def combine_requests_with_affinity(
+        rs: List[ResourceRequest],
```
Comment (jjyao) [inline python/ray/autoscaler/v2/utils.py:203]: rs?

File: python/ray/autoscaler/v2/utils.py [hunk 6]
```diff
@@ -157,8 +172,96 @@ def make(resources_map: Dict[str, float]) -> ResourceRequest:
...
+            rs: The list of resource requests to be combined.
+        Returns:
+            A list of combined resource requests.
+        """
+
+        # Map of set of serialized affinity constraint to the list of resource requests
+        requests_by_affinity: Dict[str, List[ResourceRequest]] = defaultdict(list)
+        combined_requests: List[ResourceRequest] = []
+
+        for r in rs:
```
Comment (jjyao) [inline python/ray/autoscaler/v2/utils.py:228]: r?

File: python/ray/autoscaler/v2/utils.py [hunk 7]
```diff
@@ -157,8 +172,96 @@ def make(resources_map: Dict[str, float]) -> ResourceRequest:
...
+            seen_placement_constraint = set()
+            for r in rs:
+                # Merge the resource bundles with the same affinity constraint.
+                for k, v in r.resources_bundle.items():
+                    combined_request.resources_bundle[k] = (
+                        combined_request.resources_bundle.get(k, 0) + v
+                    )
+
+                # Add and dedup the placement constraints.
+                for constraint in r.placement_constraints:
```
Comment (jjyao) [inline python/ray/autoscaler/v2/utils.py:253]: Assume placement_constraints size is 1.

**Review by jjyao — APPROVED** [review]

**Pull request merged**